sdk/modelsrepository/azure-iot-modelsrepository/src/main/java/com/azure/iot/modelsrepository/DtmiConventions.java

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.iot.modelsrepository;
+
+import com.azure.core.util.UrlBuilder;
+import com.azure.iot.modelsrepository.implementation.ModelsRepositoryConstants;
+import com.azure.iot.modelsrepository.implementation.StandardStrings;
+
+import java.io.File;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Locale;
+import java.util.regex.Pattern;
+
+/**
+ * DtmiConventions implements the core aspects of the IoT model repo conventions
+ * which includes DTMI validation and calculating a URI path from a DTMI.
+ */
+public final class DtmiConventions {
+
+    private DtmiConventions() { }
+
+    /**
+     * A DTMI has three components: scheme, path, and version.
+     * Scheme and path are separated by a colon. Path and version are separated by a semicolon i.e. scheme : path ; version.
+     * The scheme is the string literal "dtmi" in lowercase. The path is a sequence of one or more segments, separated by colons.
+     * The version is a sequence of one or more digits. Each path segment is a non-empty string containing only letters, digits, and undersc
+     * The first character may not be a digit, and the last character may not be an underscore.
+     * The version length is limited to nine digits, because the number 999,999,999 fits in a 32-bit signed integer value.
+     * The first digit may not be zero, so there is no ambiguity regarding whether version 1 matches version 01 since the latter is invalid.
+     */
+    private static final Pattern VALID_DTMI_PATTERN = Pattern.compile("^dtmi:[A-Za-z](?:[A-Za-z0-9_]*[A-Za-z0-9])?(?::[A-Za-z](?:[A-Za-z0-9_]*[A-Za-z0-9])?)*;[1-9][0-9]{0,8}$");
+
+    /**
+     * Indicates whether a given string DTMI value is well-formed.
+     *
+     * @param dtmi DigitalTwin Model Id
+     * @return True if dtmi has a valid format. False otherwise.
+     */
+    public static boolean isValidDtmi(String dtmi) {
+        if (dtmi == null || dtmi.isEmpty()) {
+            return false;
+        }
+
+        return VALID_DTMI_PATTERN.matcher(dtmi).find();
+    }
+
+    /**
+     * Generates the model URI.
+     *
+     * @param dtmi DigitalTwin Model Id.
+     * @param repositoryUri The repository uri
+     * @param expanded Is model from precomputed values
+     * @return The model uri
```

Comment:
Should document that this can throw an `IllegalArgumentException` if the DTMI is not valid